src/Chapter3.hs

```diff
@@ -480,6 +516,50 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle
+  = NoCastle
+  | Castle String
+  deriving (Show)
+
+data Wall
+  = NoWall
+  | Wall
+  deriving (Show)
+
+data AdvancedBuilding
+  = Church
+  | Library
+  deriving (Show)
+
+data House = House
+  { people :: Int }
+  deriving (Show)
+
+mkHouse :: Int -> House
+mkHouse x
+  | 1 <= x && x <= 4 = House { people = x }
+  | otherwise = House { people = (max 1 (min 4 x)) }
```

Comment:
Interesting approach. It is usually called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `mkHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂